### PR TITLE
Adding UI to the profile model

### LINKED ISSUE #107: Missing profile UI
1)add the Register page
2)add profile details
3)edit my profile

### PR DESCRIPTION
**Pull request by Davidco94**
1. Adding a register (sign up) option.
2. update logged in profile
3. show static profile
fixes #107 
#register:
static3
#my profile:
static
#another user:
static2
#edit profile:
static3

### CODE REVIEW

**Comment by Omeramsc** [comment]
Reviewing almost 900 line changes will take more time than what we have left. can you split the PR?

**Comment by Davidco94** [comment]
I'm really sorry it came out like this, but most of the lines are about HTML and CSS files and because it's a job I'm still working on from the previous demo it accumulated like this.
Unfortunately, I won't have time now to split everything and re-upload it so that it will be ready in time for the next demo. I would appreciate your understanding "before the line of judgment"

**Comment by Omeramsc** [comment]
> I'm really sorry it came out like this, but most of the lines are about HTML and CSS files and because it's a job I'm still working on from the previous demo it accumulated like this. Unfortunately, I won't have time now to split everything and re-upload it so that it will be ready in time for the next demo. I would appreciate your understanding "before the line of judgment"

You don't have to be sorry, just be aware it **won't** be ready for the final demo. this is too much to review in 1.5 days.

**Comment by Davidco94** [comment]
What can be done to make it work? Is there a way to split it efficiently?

**Comment by Omeramsc** [comment]
> What can be done to make it work? Is there a way to split it efficiently?

Think MVP, what is `must-have` and what is `nice-to-have`? currently you're trying to implement registration and everything that relates or get affected by it. minimize it to the most basic requirement for registartion

File: users/templates/users/profile_details.html [hunk 1]
```diff
@@ -0,0 +1,160 @@
+{% extends 'main/loggedInPage.html' %}
+{% block main_content %}
+{% load static %}
+
+<div class="container">
+    <div class="main-body">
+    
+          <!-- Breadcrumb -->
```
Comment (natifridman) [inline users/templates/users/profile_details.html:8]: Any reason for this comment?

File: users/templates/users/profile_details.html [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+          <nav aria-label="breadcrumb" class="main-breadcrumb">
+            <ol class="breadcrumb">
+              <li class="breadcrumb-item"><a href="/">Home</a></li>
+              <li class="breadcrumb-item"><a href="javascript:void(0)">User</a></li>
+              <li class="breadcrumb-item active" aria-current="page">User Profile</li>
+            </ol>
+          </nav>
+          <!-- /Breadcrumb -->
```
Comment (natifridman) [inline users/templates/users/profile_details.html:16]: same

File: users/templates/users/profile_details.html [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+                        <a class="btn btn-info " target="__blank" href="edit-profile/">Edit</a>
+                      {% endif %}
+                    </div>
+                  </div>
+                </div>
+              </div>
+            </div>
+          </div>
+        </div>
+  <!-- {% if object.user == request.user %} -->
```
Comment (natifridman) [inline users/templates/users/profile_details.html:141]: same

File: users/templates/users/profile_details.html [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+      <div class="articles d-flex justify-content-start">
+        <br>
+        {% for course in courses %}
+          <div class="item">
+            {% include "one_course.html" %}
+          </div>
+        {% endfor %}
+      </div>
+    </div>
+  <!-- {% endif %} -->
```
Comment (natifridman) [inline users/templates/users/profile_details.html:157]: same

File: users/tests.py
Comment (natifridman) [inline users/tests.py:1]: Please split the tests into multiple files. It's hard to follow which tests are for the model and the views

**Review by natifridman — CHANGES_REQUESTED** [review]

File: users/templates/users/profile_details.html (hunk 1 above)
Comment (Davidco94) [reply users/templates/users/profile_details.html]: deleted

File: users/templates/users/profile_details.html (hunk 2 above)
Comment (Davidco94) [reply users/templates/users/profile_details.html]: deleted

File: users/templates/users/profile_details.html (hunk 3 above)
Comment (Davidco94) [reply users/templates/users/profile_details.html]: deleted

File: users/templates/users/profile_details.html (hunk 4 above)
Comment (Davidco94) [reply users/templates/users/profile_details.html]: deleted

File: users/tests.py
Comment (Davidco94) [reply users/tests.py:1]: done

**Review by natifridman — CHANGES_REQUESTED** [review]

File: users/templates/users/register.html [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+          <div class="card-body p-md-5">
+            <div class="row justify-content-center">
+              <div class="col-md-20 col-lg-10 col-xl-10 order-2 order-lg-1">
+                <p class="text-center h1 fw-bold mb-5 mx-1 mx-md-4 mt-4">Sign up</p>
+
+                <form class="mx-1 mx-md-4" method="post" enctype="multipart/form-data">
+
+                  {% if user_form.errors %}
+                  {% for field in user_form %}
+                      {% for error in field.errors %}
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply users/templates/users/register.html]: deleted

File: users/templates/users/register.html [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                  {% for error in user_form.non_field_errors %}
+                      <div class="alert alert-danger">
+                          <strong>{{ error|escape }}</strong>
+                      </div>
+                  {% endfor %}
+                  {% endif %}
+
+                  {% if profile_form.errors %}
+                  {% for field in profile_form %}
+                      {% for error in field.errors %}
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply users/templates/users/register.html]: deleted

File: users/tests_views.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+            meeting_method='B',
+        )
+        self.url = reverse('profile-detail', kwargs={'slug': self.user.username})
+
+    def test_profile_detail_view(self):
+        response = self.client.get(self.url)
+        self.assertEqual(response.status_code, 302)
+
+
+class EditProfileViewTest(TestCase):
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply users/tests_views.py]: deleted

File: users/tests_views.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    def test_register_views_respone(self, client, valid_registration_data):
+        response = client.post('/register/', valid_registration_data)
+        assert response.status_code == 302
+        assert response.url == '/'
+
+    def test_invalid_login_user(self, client):
+        assert not client.login(username="Not_valid", password="Not_valid")
+
+
+class ProfileModelTest(TestCase):
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply users/tests_views.py]: deleted

File: users/tests_views.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+        self.url = reverse('profile-detail', kwargs={'slug': self.user.username})
+
+    def test_profile_detail_view(self):
+        response = self.client.get(self.url)
+        self.assertEqual(response.status_code, 302)
+
+
+class EditProfileViewTest(TestCase):
+    def setUp(self):
+        self.user = User.objects.create_user(
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply users/tests_views.py]: deleted

File: users/tests_views.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    def test_invalid_login_user(self, client):
+        assert not client.login(username="Not_valid", password="Not_valid")
+
+
+class ProfileModelTest(TestCase):
+    def setUp(self):
+        self.user = User.objects.create_user(
+            username='testuser',
+            password='testpassword'
+        )
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply users/tests_views.py]: deleted

File: users/tests_views.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+        self.assertEqual(self.profile.account_type, 'S')
+        self.assertEqual(self.profile.bio, 'Test bio')
+        self.assertEqual(self.profile.profession, 'Test profession')
+        self.assertEqual(self.profile.phone_number, '1234567890')
+        self.assertEqual(self.profile.city, 'Test City')
+        self.assertEqual(self.profile.meeting_method, 'B')
+
+
+class ProfileDetailViewTest(TestCase):
+    def setUp(self):
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply users/tests_views.py]: deleted

File: tests/tests_views.py
Comment (natifridman) [inline tests/tests_views.py:1]: Since this file is in the main tests folder, its name should be self-explained. So something like `tests_user_views`

File: tests/tests_views.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        assert 'profile_details.html' in course_response.templates[0].name
+
+    def test_profile_page_unauthorized(self, client, persist_user_profile):
+        course_response = client.get('/users/'+persist_user_profile.user.username, follow=True)
+        assert course_response.status_code == 200
+        assert 'main/login.html' in course_response.templates[0].name
+
+
+@pytest.mark.django_db
+class TestRegester:
```
Comment (natifridman) [inline tests/tests_views.py:76]: Typo in register, what is this class checking that the view class isn't?

File: tests/tests_views.py [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+    def test_regester_page(self, client):
```
Comment (natifridman) [inline tests/tests_views.py:78]: typo

File: tests/tests_views.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+        profile_response = client.get('/register', follow=True)
+        assert profile_response.status_code == 200
+        assert 'register.html' in profile_response.templates[0].name
+
+    def test_RegistrationForm_valid_save(self, valid_registration_data):
```
Comment (natifridman) [inline tests/tests_views.py:83]: Please follow PEP8 for naming conventions. Also, why do you need to check this form like that? sending the data on the page isn't enough?

File: tests/tests_views.py [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+        form = RegistrationForm(data=valid_registration_data)
+        assert form.is_valid()
+        user = form.save()
+        assert user is not None
+
+    def test_ProfileForm_valid_save(self, valid_profile_update_data):
```
Comment (natifridman) [inline tests/tests_views.py:89]: Please follow PEP8 for naming conventions. Also, why do you need to check this form like that? sending the data on the page isn't enough?

File: tests/tests_views.py [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+        assert 'users/register.html' in [template.name for template in response.templates]
+
+    def test_edit_profile_view_unauthenticated(self, client):
+        user = User.objects.create_user(username='testuser', password='testpassword')
+        url = reverse('users-edit-profile', kwargs={'slug': user.username})
+        response = client.get(url)
+        assert response.status_code == 302
+
+    def test_edit_profile_view_authenticated(self):
+        client = Client()
```
Comment (natifridman) [inline tests/tests_views.py:115]: Please use the client fixture

File: tests/tests_views.py [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+        user = User.objects.create_user(username='testuser', password='testpassword')
+        Profile.objects.create(user=user)
+        client.login(username='testuser', password='testpassword')
+        url = reverse('users-edit-profile', kwargs={'slug': user.username})
+        response = client.get(url)
+        assert response.status_code == 200
+        assert 'users/edit-profile.html' in [template.name for template in response.templates]
+
+    def test_profile_detail_view(self, client):
+        user = User.objects.create_user(username='testuser', password='testpassword')
```
Comment (natifridman) [inline tests/tests_views.py:125]: Please use fixture here

File: tests/tests_views.py [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+
+    def test_register_views_respone(self, client, valid_registration_data):
+        response = client.post('/register/', valid_registration_data)
+        assert response.status_code == 302
+        assert response.url == '/'
+
+    def test_invalid_login_user(self, client):
+        assert not client.login(username="Not_valid", password="Not_valid")
+
+    def test_UserUpdateForm_valid_save(self):
```
Comment (natifridman) [inline tests/tests_views.py:139]: Why this test?

File: tests/tests_views.py [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+            'email': 'new_test_email@test.com',
+            'first_name': 'Test',
+            'last_name': 'User'
+        }
+        form = UserUpdateForm(data=form_data, instance=test_user)
+        assert form.is_valid()
+        user = form.save()
+        assert user.username == 'new_test_username'
+
+    def test_ProfileUpdateForm_valid_save(self, valid_profile_update_data):
```
Comment (natifridman) [inline tests/tests_views.py:152]: same

**Review by natifridman — CHANGES_REQUESTED** [review]

**Comment by natifridman** [comment]
Please fix the typo in the PR title

File: tests/tests_views.py
Comment (Davidco94) [reply tests/tests_views.py:1]: done

File: tests/tests_views.py (hunk 12 above)
Comment (Davidco94) [reply tests/tests_views.py:76]: I connected the classes

File: tests/tests_views.py (hunk 13 above)
Comment (Davidco94) [reply tests/tests_views.py:78]: done

File: tests/tests_views.py (hunk 14 above)
Comment (Davidco94) [reply tests/tests_views.py:83]: deleted

File: tests/tests_views.py (hunk 15 above)
Comment (Davidco94) [reply tests/tests_views.py:89]: deleted

File: tests/tests_views.py (hunk 16 above)
Comment (Davidco94) [reply tests/tests_views.py:115]: done

File: tests/tests_views.py (hunk 17 above)
Comment (Davidco94) [reply tests/tests_views.py:125]: done

File: tests/tests_views.py (hunk 18 above)
Comment (Davidco94) [reply tests/tests_views.py:139]: deleted

File: tests/tests_views.py (hunk 19 above)
Comment (Davidco94) [reply tests/tests_views.py:152]: deleted

File: users/tests_views.py [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+        'phone_number': '+1234567890',
+        'city': 'Test City',
+        'account_type': Account_type.TEACHER,
+        'meeting_method': Meeting_method.LIVE
+    }
+
+
+@pytest.mark.django_db
+class TestForms:
+
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply users/tests_views.py]: deleted

File: tests/tests_views.py
Comment (natifridman) [reply tests/tests_views.py:1]: Looks the same

File: tests/tests_views.py (hunk 13 above)
Comment (natifridman) [reply tests/tests_views.py:78]: Looks the same

File: tests/tests_views.py (hunk 14 above)
Comment (natifridman) [reply tests/tests_views.py:83]: Looks the same

File: tests/tests_views.py (hunk 15 above)
Comment (natifridman) [reply tests/tests_views.py:89]: Looks the same

**Review by natifridman — CHANGES_REQUESTED** [review]
Please push the changes, it looks the same as in the last review

**Review by natifridman — CHANGES_REQUESTED** [review]
Please push the changes, it looks the same as in the last review

File: tests/tests_views.py (hunk 15 above)
Comment (Davidco94) [reply tests/tests_views.py]: now it's ok

File: tests/tests_views.py
Comment (Davidco94) [reply tests/tests_views.py:1]: I forgot to do a push after the rebase

**Review by natifridman — CHANGES_REQUESTED** [review]
I think we can add some more tests. Where do you check the update functionality for example?

**Review by natifridman — CHANGES_REQUESTED** [review]
Please resolve conflicts

File: tests/tests_user_views.py [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+        course_response = client.get('/users/'+persist_user_profile.user.username, follow=True)
+        assert course_response.status_code == 200
+        assert 'main/login.html' in course_response.templates[0].name
+
+    def test_register_page(self, client):
+        profile_response = client.get('/register', follow=True)
+        assert profile_response.status_code == 200
+        assert 'register.html' in profile_response.templates[0].name
+
+    def test_register_view(self, client):
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply tests/tests_user_views.py:93]: change name

File: tests/tests_user_views.py [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+        response = client.get(reverse('signup'))
+        assert response.status_code == 200
+        assert 'users/register.html' in [template.name for template in response.templates]
+
+    def test_edit_profile_view_unauthenticated(self, client, create_test_user):
+        user = create_test_user
+        url = reverse('users-edit-profile', kwargs={'slug': user.username})
+        response = client.get(url)
+        assert response.status_code == 302
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply tests/tests_user_views.py]: done

File: tests/tests_user_views.py [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+
+    def test_edit_profile_view_authenticated(self, client, create_test_user):
+        user = create_test_user
+        Profile.objects.create(user=user)
+        client.login(username='testuser', password='testpassword')
+        url = reverse('users-edit-profile', kwargs={'slug': user.username})
+        response = client.get(url)
+        assert response.status_code == 302
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply tests/tests_user_views.py]: done

File: tests/tests_user_views.py [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+
+    def test_edit_profile_view_unauthenticated(self, client, create_test_user):
+        user = create_test_user
+        url = reverse('users-edit-profile', kwargs={'slug': user.username})
+        response = client.get(url)
+        assert response.status_code == 302
+
+    def test_edit_profile_view_authenticated(self, client, create_test_user):
+        user = create_test_user
+        Profile.objects.create(user=user)
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply tests/tests_user_views.py]: done

File: tests/tests_user_views.py [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+        client.login(username='testuser', password='testpassword')
+        url = reverse('users-edit-profile', kwargs={'slug': user.username})
+        response = client.get(url)
+        assert response.status_code == 302
+
+    def test_profile_detail_view(self, client, create_test_user):
+        user = create_test_user
+        profile = Profile.objects.create(user=user)
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply tests/tests_user_views.py]: done

File: tests/tests_user_views.py [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+        url = reverse('profile-detail', args=[profile.user.username])
+        response = client.get(url)
+        assert response.status_code == 302
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply tests/tests_user_views.py]: done

File: tests/tests_user_views.py [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+
+    def test_register_views_respone(self, client, valid_registration_data):
+        response = client.post('/register/', valid_registration_data)
+        assert response.status_code == 302
+        assert response.url == '/'
+
+    def test_invalid_login_user(self, client):
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply tests/tests_user_views.py]: deleted

File: tests/tests_user_views.py [hunk 28]
```diff
@@ -0,0 +1,98 @@
...
+        assert response.status_code == 302
+
+    def test_profile_detail_view(self, client, create_test_user):
+        user = create_test_user
+        profile = Profile.objects.create(user=user)
+        url = reverse('profile-detail', args=[profile.user.username])
+        response = client.get(url)
+        assert response.status_code == 302
+
+    def test_register_views_respone(self, client, valid_registration_data):
```
[the remark this answers is not in the document]
Comment (Davidco94) [reply tests/tests_user_views.py]: change name

**Review by shirazyomtov — APPROVED** [review]
LGTM

**Review by shirazyomtov — APPROVED** [review]
LGTM

**Review by natifridman — CHANGES_REQUESTED** [review]

**Comment by natifridman** [comment]
> Hi Nati, I have a problem with the tests because I work with two forms - user and profile, I am having trouble sending POST requests in the URL so that the forms will be valid..what can I do with this problem?

You shouldn't check the form itself. You need to check the UI, so you should send HTTP request with the relevant data

File: tests/tests_user_views.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (Davidco94) [reply tests/tests_user_views.py:110]: deleted

**Comment by Davidco94** [comment]
Thank you very much it worked!

File: tests/tests_user_views.py [hunk 29]
```diff
@@ -0,0 +1,124 @@
...
+
+    def test_users_update_view_loaded_authorized(self, authorized_client, persist_user_profile):
+        user_detail_response = authorized_client.get(f"/users/{persist_user_profile.pk}/edit-profile/")
+        assert user_detail_response.status_code == 200
+        assert 'edit-profile.html' in user_detail_response.templates[0].name
+
+    @pytest.mark.parametrize("email, city", [(EMAIL, "city_test")])
+    def test_update_user(self, persist_user_profile, authorized_client, email, city):
+
+        profile = Profile.objects.filter(user=persist_user_profile.user).first()
```
Comment (natifridman) [inline tests/tests_user_views.py:82]: Please use a fixture for this

File: tests/tests_user_views.py [hunk 30]
```diff
@@ -0,0 +1,124 @@
...
+        assert 'main/login.html' in users_detail_response.templates[0].name
+
+    def test_edit_profile_view_unauthenticated(self, client, new_profile):
+        url = reverse('users-edit-profile', kwargs={'slug': new_profile.user.username})
+        response = client.get(url, follow=True)
+        assert response.status_code == 200
+        assert 'main/login.html' in response.templates[0].name
+
+    def test_edit_profile_view_authenticated(self, client, new_profile):
+        client.login(username='testuser', password='testpassword')
```
Comment (natifridman) [inline tests/tests_user_views.py:68]: Please use a fixture for this

File: tests/tests_user_views.py [hunk 31]
```diff
@@ -0,0 +1,124 @@
...
+                       "bio": "",
+                       "profession": "",
+                       "city": city,
+                       "phone_number": '0500000000',
+                       "account_type": 'B',
+                       "meeting_method": "B"}
+
+        response = authorized_client.post(f"/users/{persist_user_profile.user.username}/edit-profile/",
+                                          new_profile, follow=True)
+        profile = Profile.objects.filter(user=persist_user_profile.user).first()
```
Comment (natifridman) [inline tests/tests_user_views.py:99]: I don't think this is needed, you already test this part in the model

File: tests/tests_user_views.py [hunk 32]
```diff
@@ -0,0 +1,124 @@
...
+        assert response.status_code == 200
+        assert profile.city == city
+        assert profile.user.email == email
```
Comment (natifridman) [inline tests/tests_user_views.py:102]: I would check the content instead

**Review by natifridman — CHANGES_REQUESTED** [review]

File: tests/tests_user_views.py (hunk 29 above)
Comment (Davidco94) [reply tests/tests_user_views.py:82]: There is a problem with the persist_user_profile fixture, it doesn't let you update fields and that's the only way it worked

File: tests/tests_user_views.py (hunk 32 above)
Comment (Davidco94) [reply tests/tests_user_views.py:102]: 'יWhat does it mean to check the content?

File: tests/tests_user_views.py (hunk 31 above)
Comment (Davidco94) [reply tests/tests_user_views.py:99]: but the test is failed without this line

File: tests/tests_user_views.py (hunk 32 above)
Comment (natifridman) [reply tests/tests_user_views.py:102]: Something like [this](https://github.com/redhat-beyond/HighTeach/blob/d8a0662266c944e47d87dd52fa2275b0bb894176/tests/course/test_review_view.py#L28)

File: tests/tests_user_views.py (hunk 31 above)
Comment (natifridman) [reply tests/tests_user_views.py:99]: Well, you can't delete just this line... see my next comment

File: tests/tests_user_views.py (hunk 30 above)
Comment (Davidco94) [reply tests/tests_user_views.py]: done

File: tests/tests_user_views.py (hunk 32 above)
Comment (Davidco94) [reply tests/tests_user_views.py:100]: After making posts, you get a redirection to the home page and there is nothing to check in terms of the content of the response.. Isn't it better to check in the data base that the values are indeed updated?

File: tests/tests_user_views.py (hunk 31 above)
Comment (Davidco94) [reply tests/tests_user_views.py:97]: The persistent user is not updated automatically, so to check for an update I have to actually pull it out from the database again, that's why this line is necessary

**Review by natifridman — APPROVED** [review]
